src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
+            size++;
+        } else {
+            Node<T> indexNode = getNode(index);
+            Node<T> current = new Node<>(indexNode.prev, value, indexNode);
+            current.next.prev = current;
+            current.prev.next = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T replacedValue = getNode(index).value;
+        getNode(index).value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNode(index);
+        T removedValue = currentNode.value;
+        unlink(currentNode);
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (isEqual(current.value, object)) {
+                unlink(getNode(i));
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        private Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> current = head;
+        if ((size / 2) > index) {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size; i > index + 1; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(index
+                    + " is invalid index for " + size);
+        }
+    }
+
+    private void unlink(Node<T> currentNode) {
+        if (size == 1) {
+            makeEmpty();
+            return;
+        }
+        if (currentNode.next == null) {
+            tail = currentNode.prev;
+            currentNode.prev.next = null;
+        } else {
+            currentNode.next.prev = currentNode.prev;
+        }
+        if (currentNode.prev == null) {
+            head = currentNode.next;
+        } else {
+            currentNode.prev.next = currentNode.next;
+        }
```

Comment:
```suggestion
        if (currentNode.next == null) {
            tail = currentNode.prev;
            currentNode.prev.next = null;
        } else if (currentNode.prev == null) {
            head = currentNode.next;
        } else {
            currentNode.next.prev = currentNode.prev;
            currentNode.prev.next = currentNode.next;
        }
```

Reply:
Damn... Why is it so obvious and so hard to see at the same time))